msm/src/mvlookup.rs

```diff
@@ -121,15 +130,154 @@ impl<'lt, G, ID: LookupTableID> IntoIterator for &'lt LookupProof<G, ID> {
+    lookups: Vec<MVLookup<E<F>, ID>>,
+) -> E<F> {
+    let joint_combiner = {
+        let joint_combiner = ConstantExpr::from(ChallengeTerm::JointCombiner);
+        E::Atom(ExprInner::Constant(joint_combiner))
+    };
+    let beta = {
+        let beta = ConstantExpr::from(ChallengeTerm::Beta);
+        E::Atom(ExprInner::Constant(beta))
+    };
+
+    // Compute (β + f_{i}(X)) for each i.
+    // Note that f_i(X) = table_id + r * x_{1} + r^2 x_{2} + ... r^{N} x_{N}
+    let denominators = lookups
+        .iter()
+        .map(|x| {
+            // Compute r * x_{1} + r^2 x_{2} + ... r^{N} x_{N}
+            let combined_value = x
+                .value
+                .iter()
+                .rev()
+                .fold(E::zero(), |acc, y| acc * joint_combiner.clone() + y.clone())
+                * joint_combiner.clone();
+            // FIXME: sanity check for the domain, we should consider it in prover.rs.
+            // We do only support degree one constraint in the denominator.
+            assert_eq!(combined_value.degree(1, 0), 1, "Only degree one is supported in the denominator of the lookup because of the maximum degree supported (8)");
+            // add table id + evaluation point
+            beta.clone() + combined_value + x.table_id.to_constraint()
+        })
+        .collect::<Vec<_>>();
+    // Compute `column * (\prod_{i = 1}^{N} (β + f_{i}(X)))`
+    let lhs = denominators
+        .iter()
+        .fold(curr_cell(column), |acc, x| acc * x.clone());
+    let rhs = lookups
+        .into_iter()
+        .enumerate()
+        .map(|(i, x)| {
+            denominators.iter().enumerate().fold(
+                // Compute individual \sum_{j = 1, j \neq i}^{N} (β + f_{j}(X))
+                // This is the inner part of rhs. It multiplies with m_{i}
+                x.numerator,
+                |acc, (j, y)| {
+                    if i == j {
+                        acc
+                    } else {
+                        acc * y.clone()
+                    }
+                },
+            )
+        })
+        // Individual sums
+        .reduce(|x, y| x + y)
+        .unwrap_or(E::zero());
+    lhs - rhs
+}
+
+/// Build the constraints for the lookup protocol.
+/// The constraints are the partial sum and the aggregation of the partial sums.
+pub fn constraint_lookups<F: PrimeField, ID: LookupTableID>(
+    lookups_map: &BTreeMap<ID, Vec<MVLookup<E<F>, ID>>>,
+) -> Vec<E<F>> {
+    let mut constraints: Vec<E<F>> = vec![];
+    let mut idx_partial_sum = 0;
+    lookups_map.iter().for_each(|(id, lookups)| {
+        let table_lookup = MVLookup {
+            table_id: *id,
+            numerator: curr_cell(Column::LookupMultiplicity(id.to_u32())),
+            value: vec![curr_cell(Column::LookupFixedTable(id.to_u32()))],
+        };
+        // FIXME: do not clone
+        let mut lookups = lookups.clone();
+        lookups.push(table_lookup);
+        // We split in chunks of 6 (MAX_SUPPORTED_DEGREE - 2)
```

Comment:
I believe Matthew told me something about the inverses and the fact that we might not want to split in chunks. Did he probably refer just to the implementation in folding?